### PR TITLE
[Ledger] Propagate result from transaction execution to Tx status cache

### LINKED ISSUE #900: Generalise smart contract return types to handle any type (OR restrict smart contract returns to int)
Possibly an open question about the return types from executing a smart contract. At one point Ed suggested effectively an int error code

### PR DESCRIPTION
**Pull request by pbukva**
Resolves #900

### CODE REVIEW

File: libs/ledger/include/ledger/execution_result.hpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  RESOURCE_FAILURE,
+  INEXPLICABLE_FAILURE,
+};
+
+struct ContractExecutionResult
+{
+  ContractExecutionStatus status{
+      ContractExecutionStatus::NOT_RUN};  ///< The status of the transaction
+  TokenAmount charge{0};                  ///< The number of units of charge that
+  TokenAmount charge_rate{0};             ///< The cost of each unit of charge
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/execution_result.hpp:53]: Would also be good to include the charge limit here also

File: libs/ledger/include/ledger/transaction_status_cache.hpp [hunk 2]
```diff
@@ -20,14 +20,15 @@
...
+#include "ledger/execution_result.hpp"
 
 #include <chrono>
 #include <unordered_map>
 
 namespace fetch {
 namespace ledger {
 
-enum class TransactionStatus
+enum class eTransactionStatus : uint8_t
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/transaction_status_cache.hpp]: Style 🚓 - Please revert to original name

File: libs/ledger/include/ledger/chaincode/contract.hpp [hunk 3]
```diff
@@ -48,22 +48,28 @@ class Address;
 class Contract
 {
 public:
-  enum class Status
+  enum class eStatus
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/chaincode/contract.hpp]: Style  🚓 - Please revert to original name

File: libs/ledger/src/tx_status_http_interface.cpp [hunk 4]
```diff
@@ -35,6 +35,25 @@ using fetch::variant::Variant;
 namespace fetch {
 namespace ledger {
 
+namespace {
+Variant toVariant(Digest const &digest, TransactionStatusCache::TxStatus const &tx_status)
```
Comment (ejfitzgerald) [inline libs/ledger/src/tx_status_http_interface.cpp]: Style 🚓 - `ToVariant`

File: libs/ledger/src/tx_status_http_interface.cpp [hunk 5]
```diff
@@ -35,6 +35,25 @@ using fetch::variant::Variant;
...
+{
+  auto retval{Variant::Object()};
+  retval["tx"]     = ToBase64(digest);
+  retval["status"] = ToString(tx_status.status);
+  auto contr_ex_res_v{Variant::Object()};
```
Comment (ejfitzgerald) [inline libs/ledger/src/tx_status_http_interface.cpp]: Not sure if we need this extra object level here. I think it makes more sense to have a single level as a return value

File: libs/ledger/src/transaction_status_cache.cpp [hunk 6]
```diff
@@ -27,54 +27,55 @@ using fetch::generics::MilliTimer;
...
+      status = it->second;
     }
   }
 
   return status;
 }
 
-void TransactionStatusCache::Update(Digest digest, TransactionStatus status, Timepoint const &now)
+void TransactionStatusCache::Update(Digest digest, eTransactionStatus status,
+                                    ContractExecutionResult const execRes, Timepoint const &now)
```
Comment (ejfitzgerald) [inline libs/ledger/src/transaction_status_cache.cpp]: Style 🚓  - `exec_result`?

Also I think that we want to remove the `const` in order to make the move useful

Comment (ejfitzgerald) [reply libs/ledger/src/transaction_status_cache.cpp]: I also think I would probably separate the two types of update call (high level and execution level) into two unique methods and sort of the details internally.

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
This is looking good. I would not have bothered with all the renaming of the enums.

As discussed I think it is important to create a new enum that combines that high level status and execution status which can be returned from the HTTP queries. It might make sense to simple do this mapping in at the point of the query at the moment.

**Comment by pbukva** [comment]
End-to-End test for `getBlockNumber()` is failing since the python api (`ledger-api-py` repo) has not been updated.

**Review by ejfitzgerald — DISMISSED** [review]
LGTM

**Comment by pbukva** [comment]
I'm merging this in, disregarding the failed `thread_pool_gtest` (on 677817d, PR-head @ GCC8-Release), since the test is empirically well known to be flaky, and it passed in PR-merge @ GCC8-Release.

**Pull request merged**